Cythonize DistributionAccumulator

boyuanzz opened the pull request:
Make DistirbutionAccumulator all cythonized to improve performance

r:  @pabloem Please review code changes:
[distribution_counter.pxd](https://github.com/apache/beam/pull/5028/files#diff-8cc8004b0aabd3a3a73027e022b59925) & [distribution_counter.pyx](https://github.com/apache/beam/pull/5028/files#diff-de946ea52f0492d898ecf61c8e051ee7):
Cythonize DistributionAccumulator. 
Original python implementation: https://github.com/apache/beam/pull/4613 & https://github.com/apache/beam/pull/4928

[distribution_counter_test.py](https://github.com/apache/beam/pull/5028/files#diff-60375f462100c515e690f0774535fa1c):
Unit tests for new Distributionaccumulator

[cy_combiners.py](https://github.com/apache/beam/pull/5028/files#diff-bb661f94253059d6a39b383a30fd9a8e) & [cy_combiners.pxd](https://github.com/apache/beam/pull/5028/files#diff-636b00c67c9864284b0c400e76d9be32):
Delete all original python implementation in https://github.com/apache/beam/pull/4613 & https://github.com/apache/beam/pull/4928
Added DistributionCounterFn, make new DistributionAccumulator as accumulator

[cy_combiners_test.py & distribution_counter_test.py](https://github.com/apache/beam/pull/5028/files#diff-60375f462100c515e690f0774535fa1cl):
Copy most part of DistributionAccumulator unit tests.

[counters.py](https://github.com/apache/beam/pull/5028/files#diff-3c5714762090ef2e5b6c437303593494):
Create Counter.Distribution
Make new DistributionAccumulator can work with CounterFactory

[apiclient.py](https://github.com/apache/beam/pull/5028/files#diff-44359052680b20ce4f22a9436a83c43al):
Make new DistributionAccumulator can report to DataflowService

[apiclient_test.py](https://github.com/apache/beam/pull/5028/files#diff-32b72115a725f11ea87252abe743c001l):
Unit test for apiclient.translate_distribution

[distribution_counter_microbenchmark.py](https://github.com/apache/beam/pull/5028/files#diff-85403426a425dc32d6b067a593732dab):
Make benchmarks point to new DistributionAccumulator

[generac_pydoc.sh](https://github.com/apache/beam/pull/5028/files#diff-3aab4fb0a1e6b81f82ddd7038e52cc26):
Exclude distribution_counter 

r: @robertwb Please review performance part
Performance results by running distribution_counter_microbenchmark.py
```
Cythonized DistributionAccumulator:
Per element update time cost: 3.81579399109e-08
```
```
Pure python DistributionAccumulator
Per element update time cost: 4.86792087555e-07
``` 
cc: @charlesccychen

boyuanzz commented:
Performance results by running distribution_counter_microbenchmark.py
```
Cythonized DistributionAccumulator:
Per element update time cost: 3.81579399109e-08
```
```
Pure python DistributionAccumulator
Per element update time cost: 4.86792087555e-07
```

pabloem commented on sdks/python/apache_beam/utils/counters.py:
```diff
@@ -170,7 +170,11 @@ class AccumulatorCombineFnCounter(Counter):
-    self._fast_add_input = self.accumulator.add_input
+    # DistributionAccumulator is c-access(cdef) only
+    if isinstance(combine_fn, cy_combiners.DistributionCounterFn):
+      self._fast_add_input = self.accumulator.get_add_input_fn()
```
<!--new_thread; commit:e4f1c6aff932ff009d614b61e79a103184529502; resolved:0-->
I'm curious. Why do you need this? Can't you just have `cdef _add_input` and `@property\ cpdef add_input(self): return self._add_input` or something like that?
Maybe I just misunderstood the implementation though.

robertwb replied:
I just commented on this above, but here _fast_add_input is a Python wrapper for the C function (that still has all the Python overhead). cpdef does this for you; just declare add_input as cpdef and then you don't have to change the code here.

charlesccychen commented:
As a general comment, we need the classes in cy_combiners.py to be available even if Cython is not installed.  If you look at the existing code, this is respected by providing speedups using Cython annotations in cy_combiners.pxd.  Can we do something similar here?  (If it is too hard to provide these speed-ups using a pxd file, then we can switch on a slow or fast implementation, as available.)

charlesccychen commented:
In addition, note that tests do not pass:

```
Traceback (most recent call last):
  File "apache_beam/examples/complete/autocomplete_test.py", line 22, in <module>
    import apache_beam as beam
  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_MavenInstall/src/sdks/python/target/.tox/py27/local/lib/python2.7/site-packages/apache_beam/__init__.py", line 87, in <module>
    from apache_beam import io
  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_MavenInstall/src/sdks/python/target/.tox/py27/local/lib/python2.7/site-packages/apache_beam/io/__init__.py", line 21, in <module>
    from apache_beam.io.avroio import *
  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_MavenInstall/src/sdks/python/target/.tox/py27/local/lib/python2.7/site-packages/apache_beam/io/avroio.py", line 56, in <module>
    from apache_beam.io import filebasedsink
  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_MavenInstall/src/sdks/python/target/.tox/py27/local/lib/python2.7/site-packages/apache_beam/io/filebasedsink.py", line 31, in <module>
    from apache_beam.io import iobase
  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_MavenInstall/src/sdks/python/target/.tox/py27/local/lib/python2.7/site-packages/apache_beam/io/iobase.py", line 44, in <module>
    from apache_beam.transforms import core
  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_MavenInstall/src/sdks/python/target/.tox/py27/local/lib/python2.7/site-packages/apache_beam/transforms/__init__.py", line 21, in <module>
    from apache_beam.transforms import combiners
  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_MavenInstall/src/sdks/python/target/.tox/py27/local/lib/python2.7/site-packages/apache_beam/transforms/combiners.py", line 26, in <module>
    from apache_beam.transforms import cy_combiners
  File "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_Python_MavenInstall/src/sdks/python/target/.tox/py27/local/lib/python2.7/site-packages/apache_beam/transforms/cy_combiners.py", line 27, in <module>
    from apache_beam.transforms.distribution_counter import DistributionAccumulator
ImportError: No module named distribution_counter
```

boyuanzz commented:
@charlesccychen, I will update a slow implementation, which will do nothing in next commit. And after updating, these tests should pass. Thanks for your comments!

charlesccychen commented on sdks/python/apache_beam/tools/distribution_counter_microbenchmark.py:
```diff
@@ -47,21 +46,22 @@ def run_benchmark(num_runs=100, num_input=10000, seed=time.time()):
   upper_bound = sys.maxint
   inputs = generate_input_values(num_input, lower_bound, upper_bound)
+  distribution_counter = __import__(
+      'apache_beam.transforms.distribution_counter', globals(), locals(), -1)
```
<!--new_thread; commit:53a4caa11c9ea2ff223d462eb636791164b55621; resolved:0-->
Why do we need to do this strange form of import?

boyuanzz replied:
distribution_counter can be imported only when Cython available. So import distribution_counter at the beginning may raise ImportError.

charlesccychen commented on sdks/python/apache_beam/transforms/distribution_counter.pyx:
```diff
@@ -0,0 +1,134 @@
+      histogram.bucketCounts.append(self.buckets[index])
+
+  cpdef bint add_inputs_for_test(self, elements) except -1:
+    """ONLY used for unit tests"""
```
<!--new_thread; commit:e0fdd73db41481933d3173aabe6300b65003abd6; resolved:0-->
Why not implement these in the test file, if they are only needed for tests?

boyuanzz replied:
Since test files are python codes and they cannot access to cdef function. However, we need cdef add_input() to make operation faster. Then these test_purpose_only cpdef functions will behave as a bridge to make cdef functions callable from python codes.

charlesccychen replied:
Are you saying that defining `add_input` as `cpdef` as opposed to `cdef` makes this significantly slower?

boyuanzz replied:
yes, we had a test before. cdef can reduce almost 60-70ns per element, compared to cpdef.

charlesccychen replied:
Okay, thanks for the clarification.  Can you define, for example, `cdef _fast_add_input()` and use it in the Cython code here, and also define `cpdef add_input()` which just wraps this and does `return self._fast_add_input(...)`?  Similarly for `calculate_bucket_index`.  Could you also add a comment that this pattern is done for performance?

charlesccychen replied:
No, what I mean is that you can provide an external interface with `cpdef add_input`, but you can continue using the current fast `cdef _fast_add_input` directly in the Cython code.

boyuanzz replied:
Means, do following signature changes:
```
cdef add_input -> cdef _fast_add_input
cpdef get_add_input_fn -> cpdef add_input
```
with the same implementations ?

boyuanzz replied:
add_inputs_for_test will only be called from microbenchmark and unit tests. And we only care about performance when running microbenchmark. So performance can be calculated more precisely by calling add_inputs_for_test once, then a bunch of add_input will be called through c call.

robertwb replied:
You could put add_inputs_for_test in a compiled test file, but it seems overkill to create a whole new file just for this one function, so I think it's fine to leave it here.

charlesccychen commented on sdks/python/apache_beam/transforms/distribution_counter.pyx:
```diff
@@ -0,0 +1,134 @@
+    buckets_per_10: 3 buckets for every power of ten -> 1, 2, 5
+  """
+  def __init__(self):
+    self.min = INT64_MAX
```
<!--new_thread; commit:e0fdd73db41481933d3173aabe6300b65003abd6; resolved:0-->
Can you add cython definitions for these class properties?

boyuanzz replied:
cython definitions are in [distribution_counter.pxd](https://github.com/apache/beam/pull/5028/files#diff-8cc8004b0aabd3a3a73027e022b59925)

pabloem commented:
I don't think we'd want a slow and a fast implementation. I think we should only support this if the fast implementation is available, and not do anything otherwise.

charlesccychen commented:
@pabloem do you mean to say that distribution counters will not be available at all if Cython is not installed?

charlesccychen commented on sdks/python/apache_beam/transforms/distribution_counter.pyx:
```diff
@@ -0,0 +1,134 @@
+    self.sum = 0
+    self.first_bucket_offset = 58
+    self.last_bucket_offset = 0
+    self.buckets = <int64_t*> malloc(sizeof(int64_t) * 59)
```
Can we clean up this memory?  (Right now, it is a leak.)

boyuanzz replied:
Fixed in https://github.com/apache/beam/pull/5028/commits/38496c790993f0626d75f7afc2e54295fc80ec75

boyuanzz commented:
I think it's easier to combine DistributionAccumulator with current metric system(CounterFactory, apiclient) using slow implementation. We can have a flag to say, whether current distribution is slow version, if so, we will disable output_counter & time_counter feature. How do you feel like? @charlesccychen @pabloem

boyuanzz commented:
retest this please

charlesccychen commented:
Okay, thank you for the clarification.  My previous understanding was that the histogram structure was part of the user distribution metrics, and from your clarification, I see that this is not the case.  Please ignore that comment.

charlesccychen commented:
Okay, thank you for the clarification. My previous understanding was that the histogram structure was part of the user distribution metrics, and from your clarification, I see that this is not the case. Please ignore that comment.

As a general note on the same topic though, the naming is confusing to me, since I assumed `DistributionAccumulator` is what is used to collect data for user distribution counters.  This doesn't seem to be the case (what is the actual object type used for such user counters?), but maybe the naming can be made less confusing (this is not within the scope of the current PR).

robertwb commented:
At a high level, I think it's acceptable to make these metrics only available when Cython is available if it becomes cumbersome and expensive to implement this feature without Cython.

robertwb commented on sdks/python/apache_beam/transforms/distribution_counter.pyx:
```diff
@@ -0,0 +1,140 @@
+      bucket_offset = 2
+    return 1 + log10_floor * self.buckets_per_10 + bucket_offset
+
+  cpdef object get_add_input_fn(self):
```
Generally, it would be preferable to write `input_fn = couter.add_input` than add a `get_add_input_fn` method. However, you'll want to call add_input directly to avoid the Python call overhead (that we noted in the benchmarks). Here it creates a Python wrapper (object, note the return type) of the c method.

boyuanzz commented:
The latest commit make DataflowDistributionCounter under dataflow scope. @charlesccychen

robertwb commented on sdks/python/apache_beam/runners/dataflow/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,140 @@
+  return bit_count
+
+
+cdef int64_t get_log10_round_to_floor(int64_t element):
```
This code is a bit hard to follow. Given you have a list of powers of ten, it might make sense to just do

```
int power = 1
while element < POWER_TEN[power]:
    power += 1
return power - 1
```

robertwb commented on sdks/python/apache_beam/runners/dataflow/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,140 @@
+    cdef int64_t log10_floor = get_log10_round_to_floor(element)
+    cdef int64_t power_of_ten = POWER_TEN[log10_floor]
+    cdef int64_t bucket_offset = 0
+    if element <  power_of_ten<<1:
```
Just write 2 * power_of_ten, the compiler will optimized this.

robertwb commented on sdks/python/apache_beam/runners/dataflow/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,140 @@
+    cdef int64_t bucket_offset = 0
+    if element <  power_of_ten<<1:
+      bucket_offset = 0
+    elif element < ((power_of_ten<<2) + power_of_ten):
```
Same, write 5*, it's easier to read.

robertwb commented on sdks/python/apache_beam/runners/dataflow/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,140 @@
+    self.last_bucket_offset = 0
+    self.buckets = <int64_t*> malloc(sizeof(int64_t) * 59)
+    memset(self.buckets, 0, sizeof(int64_t) * 59)
+    self.buckets_per_10 = 3
```
Is this a constant? If so, define it above.

robertwb commented on sdks/python/apache_beam/runners/dataflow/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,140 @@
+    self.sum = 0
+    self.first_bucket_offset = 58
+    self.last_bucket_offset = 0
+    self.buckets = <int64_t*> malloc(sizeof(int64_t) * 59)
```
Where does the 59 come from?

boyuanzz replied:
We assume the max input is max_int64, which will have bucket_index as 58. Then we need 59 buckets(0-58).

robertwb commented on sdks/python/apache_beam/runners/dataflow/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,140 @@
+    self.first_bucket_offset = 58
+    self.last_bucket_offset = 0
+    self.buckets = <int64_t*> malloc(sizeof(int64_t) * 59)
+    memset(self.buckets, 0, sizeof(int64_t) * 59)
```
Perhaps just use calloc above?

robertwb commented on sdks/python/apache_beam/runners/dataflow/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,140 @@
+      histogram.bucketCounts.append(self.buckets[index])
+
+  cpdef bint add_inputs_for_test(self, elements) except -1:
+    """ONLY used for unit tests"""
```
This comment is redundant with the _for_test suffix. Instead, document why we want this.

robertwb commented on sdks/python/apache_beam/runners/dataflow/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,140 @@
+      self.add_input(element)
+
+  cpdef int64_t calculate_bucket_index_for_test(self, int64_t element):
+    """ONLY used for unit tests"""
```
Same.

boyuanzz commented:
Latest commit https://github.com/apache/beam/pull/5028/commits/b669ca918cdc042b4d116174e8229ba5a13e5c4b addressed all above comments. @robertwb please take a look at them, thanks!

boyuanzz commented:
I tried to integrate this new DistributionCounter with output_counter, and I always got circular dependency problem. I guess maybe it's better to keep DistributionCounter under transforms/ to avoid further possible circular dependency. If it's possible to make  DistributionCounter under transforms/, I'll provide python implementation as well in case someone doesn't have cython installed, but the cost is, we may have many lines of duplication code.
How do you feel like @charlesccychen @robertwb @pabloem ?

charlesccychen commented:
You can do your import in the function, right before you use it, i.e. you can put

```python
from apache_beam.runners.dataflow.py_dataflow_distribution_counter import DataflowDistributionCounter
```

right before you initialize the `DataflowDistributionCounter`.

boyuanzz commented:
Python related tests passed. Please review these changes @charlesccychen @pabloem  Thanks~

charlesccychen commented on sdks/python/apache_beam/transforms/dataflow_distribution_counter_test.py:
```diff
@@ -9,56 +9,62 @@
-Unit tests for the Distribution Counter
-'''
+"""Unit tests for DataflowDistributionCounter
+When Cyhthon is available, unit tests will test on cythonized module,
```
<!--new_thread; commit:367e48dba895159a01d9a4e098c275da83197f79; resolved:0-->
"Cython"

charlesccychen replied:
<!--thread_id:cc_180909431_t; commit:367e48dba895159a01d9a4e098c275da83197f79; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> "Cython"

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/utils/counters.py:
```diff
@@ -132,7 +132,7 @@ class Counter(object):
 
   # Distribution Accumulator Fn
-  DISTRIBUTION = cy_combiners.DistributionCounterFn()
+  DISTRIBUTION = cy_combiners.DataflowDistributionCounterFn()
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Please rename `DISTRIBUTION` to something more that doesn't imply this is used for all distribution metrics.

charlesccychen replied:
Sure.  It's also better if later we can generalize this for other cases (e.g. when Beam on Flink / portability emerges), so could you also file a JIRA for generalizing usage of this distribution counter, and reference it here in a comment?

charlesccychen commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,119 @@
+    first_bucket_offset: starting index of the first stored bucket.
+    last_bucket_offset: end index of buckets.
+    buckets: histogram buckets of value counts for a distribution
+                                                    (1,2,5 bucketing).
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Please fix weird spacing.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -359,49 +348,40 @@ def add_input(self, element):
+    self.first_bucket_offset = min(self.first_bucket_offset, bucket_index)
+    self.last_bucket_offset = max(self.last_bucket_offset, bucket_index)
 
   def calculate_bucket_index(self, element):
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Please preserve old docstring.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -359,49 +348,40 @@ def add_input(self, element):
+class DataflowDistributionCounterFn(AccumulatorCombineFn):
+  """A subclass of cy_combiners.AccumulatorCombineFn
+  Make DataflowDistributionCounter able to report to Dataflow service via
+  CounterFactory
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Add extra newline

charlesccychen commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -359,49 +348,40 @@ def add_input(self, element):
+
+
+class DataflowDistributionCounterFn(AccumulatorCombineFn):
+  """A subclass of cy_combiners.AccumulatorCombineFn
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Period after sentence, two lines after initial docstring line.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -312,44 +310,35 @@ class AllCombineFn(AccumulatorCombineFn):
-  """Distribution Counter:
-  contains value distribution statistics and methods for incrementing
+class DataflowDistributionCounter(object):
+  """Pure python DataflowDistributionCounter in case Cython not available
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Period after sentence, two lines after initial docstring line.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,119 @@
+
+
+cdef class DataflowDistributionCounter(object):
+  """Distribution Counter:
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Period after sentence, two lines after initial docstring line.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,119 @@
+      self.add_input(element)
+
+  cpdef int64_t calculate_bucket_index_for_test(self, int64_t element):
+    """Used for unit tests
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Period after sentence, two lines after initial docstring line.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,119 @@
+      histogram.bucketCounts.append(self.buckets[index])
+
+  cpdef bint add_inputs_for_test(self, elements) except -1:
+    """Used for performance microbenchmark
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Period after sentence, two lines after initial docstring line.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,119 @@
+    self.max = max(self.max, element)
+    self.count += 1
+    self.sum += element
+    cdef int64_t bucket_index = self.calculate_bucket_index(element)
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Can you rename this `_fast_calculate_bucket_index` and rename `calculate_bucket_index_for_test` to `calculate_bucket_index`?  This way, we use the fast version internally, but we expose the same interface as the non-cythonized version, so we don't have to check `is_cythonized` in this case.

boyuanzz replied:
yes but we still want to keep ```is_cythonized``` since we wanna disable counter feature if cython is not available.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,119 @@
+    buckets: histogram buckets of value counts for a distribution
+                                                    (1,2,5 bucketing).
+             max bucket_index is 58( sys.maxint as input)
+    is_cythonized: mark whether DataflowDistributionCounter cythonized
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
"whether DDataflowDistributionCounter is cythonized."; period at end of sentence.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,119 @@
+cdef class DataflowDistributionCounter(object):
+  """Distribution Counter:
+  contains value distribution statistics and methods for incrementing
+  Attributes:
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Add newline before Attributes

charlesccychen commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -359,49 +348,40 @@ def add_input(self, element):
+      bucket_offset = 2
+    return 1 + log10_floor * self.buckets_per_10 + bucket_offset
+
+  def translate_to_histogram(self, histogram):
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Please add doc, in particular with what the type of `histogram` should be.

A side note is that this is a weird pattern, where we need to pass in a preallocated `histogram`.  Can you also document why we use this pattern instead of returning a newly created `Histogram`?

boyuanzz replied:
Histogram is a class under ```runners/dataflow/internal/clients```.  According to what @robertwb suggested, it's better to avoid importing runners/dataflow into modules outside of the scope.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,119 @@
+
+  cdef int64_t calculate_bucket_index(self, int64_t element):
+    """Calculate the bucket index for the given element
+    Declare calculate_bucket_index as cdef in order to improve performance, 
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Period after sentence, two lines after initial docstring line.

boyuanzz commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,133 @@
+  cdef int64_t _fast_calculate_bucket_index(self, int64_t element):
+    """Calculate the bucket index for the given element.
+    
+    
```
Okay for this part, I guess we do have something like python style rules. Could you please attach a link here pointed to the rules we are using?

charlesccychen replied:
See https://www.python.org/dev/peps/pep-0257/

charlesccychen replied:
See https://www.python.org/dev/peps/pep-0257/

boyuanzz commented on sdks/python/apache_beam/tools/distribution_counter_microbenchmark.py:
```diff
@@ -47,21 +46,24 @@ def run_benchmark(num_runs=100, num_input=10000, seed=time.time()):
   inputs = generate_input_values(num_input, lower_bound, upper_bound)
+  distribution_counter = __import__(
+      'apache_beam.transforms.cy_dataflow_distribution_counter',
+      globals(), locals(), -1)
```
Could you please give a reason why preferring from...import rather than __import?

boyuanzz commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -359,49 +361,51 @@ def add_input(self, element):
+        self.buckets[self.first_bucket_offset:self.last_bucket_offset + 1])
+
 
+class DataflowDistributionCounterFn(AccumulatorCombineFn):
```
I guess the naming rule is like: acumullator_name+fn, eg. SumInt64Accumulator has SumInt64Fn

boyuanzz commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -312,44 +310,48 @@ class AllCombineFn(AccumulatorCombineFn):
+  return power - 1
+
+
+class DataflowDistributionCounter(object):
```
My concern is, current design of ```DataflowDistributionCounter``` is not a true Accumulator, for example,  there is no actual merge function build in(we may have it in the future if needed, but currently, I just planed to ```pass``` or ```raise NotImplementation```). 
The reason why using a ```AccumulatorCombineFn``` to wrap```DataflowDistributionCounter``` is, I need to use counter_factory to report my counter to ```DataflowService```.

boyuanzz commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -312,44 +310,48 @@ class AllCombineFn(AccumulatorCombineFn):
-    """Starting index of the first stored bucket"""
-    self.first_bucket_offset = 0
-    """There are 3 buckets for every power of ten: 1, 2, 5"""
+    self.first_bucket_offset = 58
```
<!--thread_id:cc_180949165_t; commit:89d005daa5313826869ebdad0aba047ae95febcb; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Prefer a named constant to a magic number.  Can you factor this into an integer constant and add a comment next to it describing how we came to use this value? (I know you already have this in the class docstring).
> 
> Also for the corresponding cython file.

<!--section:body-->
Done.

boyuanzz commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,133 @@
+  cdef int64_t _fast_calculate_bucket_index(self, int64_t element):
+    """Calculate the bucket index for the given element.
+    
+    
```
<!--thread_id:cc_180949166_t; commit:89d005daa5313826869ebdad0aba047ae95febcb; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> See https://www.python.org/dev/peps/pep-0257/

<!--section:body-->
Done.

boyuanzz commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,133 @@
+  cpdef int64_t calculate_bucket_index(self, int64_t element):
+    """Used for unit tests.
+    
+    
```
<!--thread_id:cc_180949161_t; commit:89d005daa5313826869ebdad0aba047ae95febcb; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Please remove extra newline (sorry, when I said two newlines, I meant one empty line in between)

<!--section:body-->
Done.

boyuanzz commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -312,44 +310,48 @@ class AllCombineFn(AccumulatorCombineFn):
-    """There are 3 buckets for every power of ten: 1, 2, 5"""
+    self.first_bucket_offset = 58
+    self.last_bucket_offset = 0
+    self.buckets = [0]*59
```
<!--thread_id:cc_180949164_t; commit:89d005daa5313826869ebdad0aba047ae95febcb; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Same as above (use `CONSTANT + 1` ?)

<!--section:body-->
Done.

boyuanzz commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -359,49 +361,51 @@ def add_input(self, element):
 
-class DistributionCounterFn(AccumulatorCombineFn):
-  _accumulator_type = DistributionAccumulator
+
```
<!--thread_id:cc_180949163_t; commit:89d005daa5313826869ebdad0aba047ae95febcb; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Please remove extra newline.

<!--section:body-->
Done.

boyuanzz commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -359,49 +361,51 @@ def add_input(self, element):
+  """
+  try:
+    from apache_beam.transforms.cy_dataflow_distribution_counter \
+      import DataflowDistributionCounter
```
<!--thread_id:cc_180949170_t; commit:89d005daa5313826869ebdad0aba047ae95febcb; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> No need to use continuation--long line for import is okay.

<!--section:body-->
Done.

boyuanzz commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,133 @@
+  cpdef bint add_inputs_for_test(self, elements) except -1:
+    """Used for performance microbenchmark.
+    
+    
```
<!--thread_id:cc_180949162_t; commit:89d005daa5313826869ebdad0aba047ae95febcb; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Please remove extra newline (sorry, when I said two newlines, I meant one empty line in between)

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -359,49 +366,49 @@ def add_input(self, element):
+  version.
+  """
+  try:
+    from apache_beam.transforms.cy_dataflow_distribution_counter import DataflowDistributionCounter
```
<!--new_thread; commit:2e80f6f09baf2949f8b46af1b2858a2fa6ee567b; resolved:0-->
See above regarding `SlowDataflowDistributionCounter`.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -312,44 +310,53 @@ class AllCombineFn(AccumulatorCombineFn):
+
+
+class DataflowDistributionCounter(object):
+  """Pure python DataflowDistributionCounter in case Cython not available.
```
<!--new_thread; commit:2e80f6f09baf2949f8b46af1b2858a2fa6ee567b; resolved:0-->
Can we rename this to `SlowDataflowDistributionCounter`?  Then, we can do:

```python
try:
  from apache_beam.transforms.cy_dataflow_distribution_counter import DataflowDistributionCounter
except:
  DistributionCounter = SlowDataflowDistributionCounter
```

Below, we can get rid of the `try` / `except` in `DataflowDistributionCounterFn` and just do `_accumulator_type = DataflowDistributionCounter`.

Further, explicitly exposing the slow counter lets us test both in tests (see the test file).

boyuanzz replied:
With current pattern in ```transforms/__init__.py```, we can do the same thing as you mentioned without renaming. For test both, I had more explainations below.

boyuanzz replied:
and for here, I feel confused about where you suggested to put this pattern?
```
try:
  from apache_beam.transforms.cy_dataflow_distribution_counter import DataflowDistributionCounter
except:
  DistributionCounter = SlowDataflowDistributionCounter
```

charlesccychen replied:
You can have that in the file directly, right after the class definitionfor example:

```python
class SlowDataflowDistributionCounter:
   ...

try:
  from apache_beam.transforms.cy_dataflow_distribution_counter import DataflowDistributionCounter
except:
  DistributionCounter = SlowDataflowDistributionCounter
```

Thank you for pointing out that we run both python-only and cython tests.  Please ignore that comment in this case.

charlesccychen commented on sdks/python/apache_beam/transforms/dataflow_distribution_counter_test.py:
```diff
@@ -9,24 +9,27 @@
+class DataflowDistributionAccumulatorTest(unittest.TestCase):
   def test_calculate_bucket_index_with_input_0(self):
-    counter = DistributionAccumulator()
+    counter = DataflowDistributionCounter()
```
<!--new_thread; commit:89d005daa5313826869ebdad0aba047ae95febcb; resolved:0-->
Can we do the following, to get extra test coverage, for each of these test methods?  Can you create a new method on this class, say, `create_distribution_counters()` that does, for example:

```python
def create_distribution_counters(self):
  counters = [SlowDataflowDistributionCounter()]
  if DataflowDistributionCounter != SlowDataflowDistributionCounter:
    # Fast Cythonized distribution counter available.
    counters.append(DataflowDistributionCounter())
  return counters
```

Then, you can wrap each of the tests here like this:

```python
for counter in self.create_distribution_counters():
  index = counter.calculate_bucket_index(0)
  self.assertEquals(index, 0)
  # ...etc.
```

This way, we get test coverage for all available distribution counter implemenations.

boyuanzz replied:
There is a import pattern in transforms/_init_.py (https://github.com/apache/beam/pull/5028/files/dd3ae5364c173a99c8a15f7e80960b9307993f7c#diff-a827371b61aef78df1df61f6d34c0716R28). So if running tox -e py27, this unit tests will test on python module, if running tox -2 py27-cython, tests will test on cython module.

charlesccychen replied:
Thank you for pointing out that we run both python-only and cython tests.  Please ignore this comment in that case.

charlesccychen commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -312,44 +310,48 @@ class AllCombineFn(AccumulatorCombineFn):
+  return power - 1
+
+
+class DataflowDistributionCounter(object):
```
<!--thread_id:cc_180949169_t; commit:89d005daa5313826869ebdad0aba047ae95febcb; resolved:1-->
<!--section:context-quote-->
> **boyuanzz** wrote:
> My concern is, current design of ```DataflowDistributionCounter``` is not a true Accumulator, for example,  there is no actual merge function build in(we may have it in the future if needed, but currently, I just planed to ```pass``` or ```raise NotImplementation```). 
> The reason why using a ```AccumulatorCombineFn``` to wrap```DataflowDistributionCounter``` is, I need to use counter_factory to report my counter to ```DataflowService```.

<!--section:body-->
Thank you for the explanation.

charlesccychen commented on sdks/python/apache_beam/runners/dataflow/internal/apiclient.py:
```diff
@@ -738,12 +739,21 @@ def to_split_int(n):
   dist_update_proto.max = to_split_int(distribution_update.max)
   dist_update_proto.count = to_split_int(distribution_update.count)
   dist_update_proto.sum = to_split_int(distribution_update.sum)
+  # Only DistributionAccumulator has buckets and first_bucket_offset.
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Please fix changed name.

boyuanzz replied:
<!--thread_id:cc_181184006_t; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Please fix changed name.

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/transforms/__init__.py:
```diff
@@ -23,3 +23,8 @@
 from apache_beam.transforms.timeutil import TimeDomain
 from apache_beam.transforms.util import *
+
+try:
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Please add comment: no backwards compatibility guarantees here.

boyuanzz replied:
<!--thread_id:cc_181184009_t; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> No, just here--this is part of the public API; I think the other places you added things already have this comment.

<!--section:body-->
Done.

boyuanzz replied:
<!--thread_id:cc_181184007_t; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Please add comment: no backwards compatibility guarantees here.

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/transforms/__init__.py:
```diff
@@ -23,3 +23,8 @@
 from apache_beam.transforms.timeutil import TimeDomain
 from apache_beam.transforms.util import *
+
+try:
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Please add comment: no backwards compatibility guarantees here.

boyuanzz replied:
So basically that is, I'm supposed to add this 'no backwards compatibility guarantees here.' in everywhere I touched, right?

charlesccychen replied:
No, just here--this is part of the public API; I think the other places you added things already have this comment.

charlesccychen commented on sdks/python/apache_beam/transforms/__init__.py:
```diff
@@ -23,3 +23,8 @@
+try:
+  from apache_beam.transforms.cy_dataflow_distribution_counter import DataflowDistributionCounter
+except ImportError:
+  from apache_beam.transforms.cy_combiners import DataflowDistributionCounter
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Is this branch obsolete?  I don't see

boyuanzz replied:
Sorry I may not 100% understand what you concerns, but the thing i want to do here is:
1. Bring DataflowDistributionCounter into transforms level
2. Try import cython module first, if failed, then import python module. Since we keep the same interfaces, I don't need to have this import pattern everywhere when using this counter.

charlesccychen replied:
Sorry, this was an incomplete comment; I made a mistake here.

charlesccychen commented on sdks/python/apache_beam/runners/dataflow/internal/apiclient_test.py:
```diff
@@ -133,6 +134,26 @@ def test_translate_distribution(self):
 
+  def test_translate_distribution_counter(self):
+    counter_update = DataflowDistributionCounter()
+    counter_update.add_input(1)
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Can you add more inputs?  With just this one input, we have that min=max=sum=count=1, which may fail to catch mistakes.

boyuanzz replied:
This test mainly focuses on testing whether ```DataflfowDistributionCounter``` can be fully converted to ```DistributionUpdate````. There is a unit test which mainly focuses on testing DistributionCounter itself.

charlesccychen replied:
Right, what I mean is that since all the values as listed above are equal to 1, we may accidentally swap fields without this test noticing.  This could be improved by simply adding, for example,  `counter_update.add_input(2)`, `counter_update.add_input(3)`.

boyuanzz replied:
Could you please explain more your idea? Current test case is designed for matching two data structure's field. I'm not very clear why we need to add more input.

charlesccychen replied:
This is really not important at all--I just meant that if, for example, we accidentally write `dist_update_proto.max = to_split_int(distribution_update.min)` where we assign the wrong field, the current test will not catch it.

charlesccychen replied:
Can you add the two lines as above? 

```
counter_update.add_input(2)
counter_update.add_input(3)
```

boyuanzz replied:
for test purpose, I think tow inputs are enough. just curious why we need three inputs here?

charlesccychen replied:
Thanks.  That's fine.

charlesccychen commented on sdks/python/apache_beam/runners/dataflow/internal/apiclient.py:
```diff
@@ -738,12 +739,21 @@ def to_split_int(n):
+  """Translate metrics DistributionUpdate to dataflow distribution update.
+  Args:
+    distribution_update: Instance of DistributionUpdate or
+    DistributionAccumulator.
```
<!--new_thread; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:0-->
Please fix changed name.

boyuanzz replied:
<!--thread_id:cc_181184014_t; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Please fix changed name.

<!--section:body-->
Done.

charlesccychen commented on sdks/python/apache_beam/utils/counters.py:
```diff
@@ -132,7 +132,7 @@ class Counter(object):
 
   # Distribution Accumulator Fn
-  DISTRIBUTION = cy_combiners.DistributionCounterFn()
+  DISTRIBUTION = cy_combiners.DataflowDistributionCounterFn()
```
<!--thread_id:cc_180909434_t; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Sure.  It's also better if later we can generalize this for other cases (e.g. when Beam on Flink / portability emerges), so could you also file a JIRA for generalizing usage of this distribution counter, and reference it here in a comment?

<!--section:body-->
Done.

boyuanzz commented on sdks/python/apache_beam/tools/distribution_counter_microbenchmark.py:
```diff
@@ -47,21 +46,24 @@ def run_benchmark(num_runs=100, num_input=10000, seed=time.time()):
   inputs = generate_input_values(num_input, lower_bound, upper_bound)
+  distribution_counter = __import__(
+      'apache_beam.transforms.cy_dataflow_distribution_counter',
+      globals(), locals(), -1)
```
<!--thread_id:cc_180949173_t; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> I think in general it is better to use more commonly used patterns when possible.  It looks like `__import__` use is somewhat discouraged according to the docs: https://docs.python.org/3/library/functions.html#__import__.
> 
> You may have to add `# pylint: disable=wrong-import-order, wrong-import-position` to get lint to pass.

<!--section:body-->
Done.

boyuanzz commented on sdks/python/apache_beam/runners/dataflow/internal/apiclient.py:
```diff
@@ -738,12 +739,21 @@ def to_split_int(n):
 
 def translate_distribution(distribution_update, metric_update_proto):
-  """Translate metrics DistributionUpdate to dataflow distribution update."""
+  """Translate metrics DistributionUpdate to dataflow distribution update.
```
<!--thread_id:cc_181184013_t; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Extra newline after.

<!--section:body-->
Done.

boyuanzz commented on sdks/python/apache_beam/transforms/__init__.py:
```diff
@@ -23,3 +23,8 @@
+try:
+  from apache_beam.transforms.cy_dataflow_distribution_counter import DataflowDistributionCounter
+except ImportError:
+  from apache_beam.transforms.cy_combiners import DataflowDistributionCounter
```
<!--thread_id:cc_181184010_t; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Sorry, this was an incomplete comment; I made a mistake here.  Please ignore.

<!--section:body-->
Done.

boyuanzz commented on sdks/python/apache_beam/runners/dataflow/internal/apiclient_test.py:
```diff
@@ -133,6 +134,26 @@ def test_translate_distribution(self):
 
+  def test_translate_distribution_counter(self):
+    counter_update = DataflowDistributionCounter()
+    counter_update.add_input(1)
```
<!--thread_id:cc_181184011_t; commit:dd3ae5364c173a99c8a15f7e80960b9307993f7c; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> This is really not important at all--I just meant that if, for example, we accidentally write `dist_update_proto.max = to_split_int(distribution_update.min)` where we assign the wrong field, the current test will not catch it.

<!--section:body-->
Done.

boyuanzz commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -312,44 +310,53 @@ class AllCombineFn(AccumulatorCombineFn):
+
+
+class DataflowDistributionCounter(object):
+  """Pure python DataflowDistributionCounter in case Cython not available.
```
<!--thread_id:cc_181183985_t; commit:2e80f6f09baf2949f8b46af1b2858a2fa6ee567b; resolved:0-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> You can have that in the file directly, right after the class definition for example:
> 
> ```python
> class SlowDataflowDistributionCounter:
>    ...
> 
> try:
>   from apache_beam.transforms.cy_dataflow_distribution_counter import DataflowDistributionCounter
> except:
>   DistributionCounter = SlowDataflowDistributionCounter
> ```
> 
> Thank you for pointing out that we run both python-only and cython tests.  Please ignore that comment in this case.

<!--section:body-->
From my understanding,  actually I'm doing the same thing now as you suggested.

charlesccychen replied:
Right, you are doing this implicitly in `DataflowDistributionCounterFn`, but you can make this more explicit by doing the swapping (if possible) right after the slow class definition, and so later you can just do:

```python
class DataflowDistributionCounterFn(AccumulatorCombineFn):
  _accumulator_type = DataflowDistributionCounter
```

boyuanzz replied:
Thanks! One more thing is, actually I don't think it's necessary to rename as Slow, do you really think it's better with Slow?

charlesccychen replied:
The advantage is that you will be able to directly refer to the Python version of this class, in the case where the Cython version replaces it.  Will leave this up to you.

boyuanzz replied:
I think we can do the same things without renaming, please refer to the latest commit https://github.com/apache/beam/pull/5028/commits/88359a8714261269d7bfa8bfdebb855db2edc7b6

charlesccychen commented on sdks/python/apache_beam/tools/distribution_counter_microbenchmark.py:
```diff
@@ -47,21 +46,22 @@ def run_benchmark(num_runs=100, num_input=10000, seed=time.time()):
   upper_bound = sys.maxint
   inputs = generate_input_values(num_input, lower_bound, upper_bound)
+  from apache_beam.transforms import DataflowDistributionCounter
   print ("Number of runs:", num_runs)
```
nit: please remove extraneous space after `print`

charlesccychen commented:
@boyuanzz: Can you please post new benchmark results with the latest version of the code?

R: @aaltay

charlesccychen commented:
Thanks Boyuan!  This is a great 20x speedup.

aaltay commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,129 @@
+    self.max = 0
+    self.count = 0
+    self.sum = 0
+    self.first_bucket_offset = MAX_BUCKET_SIZE -1
```
Missing space in `-1`. Could be fixed in a follow up PR.

@robertwb is there a way to lint cython?

robertwb replied:
We've generally written pure Python code with pxd declarations which has a side effect of nice linting, though that's harder to do. 

Some linters are tolerant of cython syntax, but most are not.

aaltay commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -312,44 +310,53 @@ class AllCombineFn(AccumulatorCombineFn):
-  y = MAX_LONG_10_FOR_LEADING_ZEROS[number_of_leading_zeros]
-  return y - compare_to(element, math.pow(10, y))
+  power = 0
+  while element >= POWER_TEN[power]:
```
Do we assume that element would be a signed integer? sys.maxint would be >= 10e18, and in that case would this loop use an out of bounds index?

boyuanzz replied:
I think sys.maxint < power(10,19)

aaltay commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,129 @@
+      bucket_offset = 1
+    else:
+      bucket_offset = 2
+    return 1 + log10_floor * BUCKET_PER_TEN + bucket_offset
```
What is the max value that could be returned here. If understand correctly it will be 1 + 18 * 3 + 2 = 57. But `MAX_BUCKET_SIZE` is 59. Not very important, should MAX_BUCKET_SIZE be 58.

robertwb replied:
I would just write MAX_BUCKET_SIZE  = bucket(sys.maxint) here rather than a constant whose value is computed manually.

boyuanzz replied:
Will we have some downsides if hold a constant value as bucket size?

aaltay commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,129 @@
+    for index in range(self.first_bucket_offset, self.last_bucket_offset+1):
+      histogram.bucketCounts.append(self.buckets[index])
+
+  cpdef bint add_inputs_for_test(self, elements) except -1:
```
nit; space after `-`

boyuanzz replied:
I don't think there should be a space after ```-```
http://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values

aaltay commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,129 @@
+    histogram.firstBucketOffset = self.first_bucket_offset
+    histogram.bucketCounts = []
+    cdef int64_t index = self.first_bucket_offset
+    for index in range(self.first_bucket_offset, self.last_bucket_offset+1):
```
nit; space around `+`

aaltay commented on sdks/python/apache_beam/transforms/dataflow_distribution_counter_test.py:
```diff
@@ -39,26 +42,23 @@ def test_calculate_bucket_index_within_max_long(self):
-    for value in [1, 500, 2, 3, 1000, 4]:
-      counter.add_input(value)
-    self.assertEquals(counter.buckets, expected_buckets)
+    for element in [1, 500, 2, 3, 1000, 4]:
```
Could you add a test for the edges? (Largest possible input etc.)

boyuanzz replied:
There is a test case https://github.com/apache/beam/pull/5028/files#diff-9e972e8dfd5893228ef8262c3f1c30fcR31 which tests calculating bucket index of max_input. I think the critical part of DistributionCounter is calculating bucket index, so will this be enough for covering edge cases?

robertwb commented on sdks/python/apache_beam/transforms/cy_combiners.py:
```diff
@@ -312,44 +310,53 @@ class AllCombineFn(AccumulatorCombineFn):
+  """Pure python DataflowDistributionCounter in case Cython not available.
+
+
+  Please avoid using python mode if possible, since it's super slow
```
<!--new_thread; commit:88359a8714261269d7bfa8bfdebb855db2edc7b6; resolved:0-->
Who can avoid this? I would be in favor of simply omitting this rather than making things way slower (plus lots of duplicated logic) in this case.

boyuanzz replied:
1. even we use python mode in cython, it's still to slow to fit in 5% budget
2. we need a pure python DistributionCounter to avoid failure when cython is not installed.
3. we cannot say to user to some extent, you must have cython installed then you can use this counter
4. I don't think we can fully control how user using open source code, we provide a faster version, but it's still possible some users only want python counter. What I can do here is to attach more running data to say how slow this python version will be then leave decision to user

robertwb commented on sdks/python/apache_beam/transforms/__init__.py:
```diff
@@ -23,3 +23,9 @@
+try:
+  from apache_beam.transforms.cy_dataflow_distribution_counter import DataflowDistributionCounter
+except ImportError:
+  from apache_beam.transforms.cy_combiners import DataflowDistributionCounter
```
<!--new_thread; commit:70a22e2b9aaf49c4f9fc6cc86ed5bf10093823e7; resolved:0-->
It's counter-intuitive that this is the one that's slow. Also, do we need this in the top -level apache_beam.transform?

boyuanzz replied:
yes, this will first trying to import cython version, if failed, then import python version. Then we can avoid a bunch of try-except clauses when importing to other modules

robertwb commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,129 @@
+  return power - 1
+
+
+cdef class DataflowDistributionCounter(object):
```
<!--new_thread; commit:2e80f6f09baf2949f8b46af1b2858a2fa6ee567b; resolved:0-->
Why is this called a DataflowDistributionCounter, rather than just DistributionCounter?

boyuanzz replied:
Since Python SDK has general distribution counter called DistributionData, this counter is designed specifically for DataflowService.

robertwb commented on sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx:
```diff
@@ -0,0 +1,129 @@
+    self.max = 0
+    self.count = 0
+    self.sum = 0
+    self.first_bucket_offset = MAX_BUCKET_SIZE -1
```
<!--new_thread; commit:2e80f6f09baf2949f8b46af1b2858a2fa6ee567b; resolved:0-->
Why is first_bucket_offset bigger than last_bucket_offset? Seems it should be the other way around. (Why do we even need to store these values?)

boyuanzz replied:
reason is, everytime a new element added in, we will use bucket index calculated based on new element to update first_index and last_index

boyuanzz replied:
I stored these values because:
1. I use Histogram in DataflowService proto reporting buckets to DataflowService, proto asks for start_bucket_index.
2. I'm supposed to report buckets from the first non-empty one to the last non-empty one, but in order to make calculation faster, I hold a fixed size bucket list, so i need a last_index to truncate bucket_list

robertwb replied:
My point was that there's no need to track them continuously when they're so easy to compute on the fly later.

boyuanzz replied:
Yes, it's very easy to do the calculation in translate_histogram func. I'm wondering, will it gain more readability if I keep these values? If not, I'll delete them.

robertwb commented on sdks/python/apache_beam/utils/counters.py:
```diff
@@ -131,8 +131,9 @@ class Counter(object):
-  DISTRIBUTION = cy_combiners.DistributionCounterFn()
+  # Dataflow Distribution Accumulator Fn.
+  # TODO(BEAM-4045): Generalize distribution counter if necessary.
+  DATAFLOW_DISTRIBUTION = cy_combiners.DataflowDistributionCounterFn()
```
<!--new_thread; commit:89d005daa5313826869ebdad0aba047ae95febcb; resolved:0-->
Isn't this using the slow one that we were trying to avoid?

boyuanzz replied:
Nope, only if cython is not available, this will be the slow one. But I'll also check whether current counter is the slow one before adding counter into worker. That's the reason why there is a field called is_cythonized

robertwb commented on sdks/python/apache_beam/transforms/dataflow_distribution_counter_test.py:
```diff
@@ -9,24 +9,27 @@
 # limitations under the License.
 #
 
```
<!--new_thread; commit:89d005daa5313826869ebdad0aba047ae95febcb; resolved:0-->
The file name doesn't seem to match what's being tested.

boyuanzz replied:
Yes, but it does match the class to test. The reason why I had this file name is because, I intented to test cython counter and python counter in one  test suite.

robertwb commented:
OK, currently we have

metricbase.Distribution
metric.DelegatingDistribution
metric.cells.DistributionCell
metric.cells.DistributionResult
metric.cells.DistributionData
metric.cells.DistributionAggregator
cy_combiners.DataflowDistributionCounter
cy_combiners.DataflowDistributionCounterFn
cy_dataflow_distribution_counter.DataflowDistributionCounter

If you hadn't seen this code before (or came back in a while) how easy would it be to tell which one to use in which case. Which ones are fast? Where are the tests? How does this relate to the code in util/counters? 

I think this all needs to be consolidated and simplified (and that's probably beyond the scope of this PR, but this PR highlights the issue. Created https://issues.apache.org/jira/browse/BEAM-4074 which should probably be accompanied by a design doc). 

For now, let's stick with the convention cy_xxx/py_xxs or xxx_fast/xxx_slow for both the modules and the classes, and move everything out of cy_combiners. I'd also drop the "dataflow" modifier, but that's up to you. 

cc @pabloem

boyuanzz commented:
Latest commit https://github.com/apache/beam/pull/5028/commits/5f54f00ae09aa2f986a15f8b58944b0dabc5f91d fixed lint issues. @aaltay

robertwb commented:
On Fri, Apr 13, 2018 at 1:50 PM Boyuan Zhang <notifications@github.com>
wrote:

> *@boyuanzz* commented on this pull request.
> ------------------------------
>
> In sdks/python/apache_beam/transforms/cy_dataflow_distribution_counter.pyx
> <https://github.com/apache/beam/pull/5028#discussion_r181506965>:
>
> > +    min: minimum value of all inputs.
> +    max: maximum value of all inputs.
> +    count: total count of all inputs.
> +    sum: sum of all inputs.
> +    first_bucket_offset: starting index of the first stored bucket.
> +    last_bucket_offset: end index of buckets.
> +    buckets: histogram buckets of value counts for a
> +    distribution(1,2,5 bucketing). Max bucket_index is 58( sys.maxint as input).
> +    is_cythonized: mark whether DataflowDistributionCounter cythonized.
> +  """
> +  def __init__(self):
> +    self.min = INT64_MAX
> +    self.max = 0
> +    self.count = 0
> +    self.sum = 0
> +    self.first_bucket_offset = MAX_BUCKET_SIZE -1
>
> Yes, it's very easy to do the calculation in translate_histogram func. I'm
> wondering, will it gain more readability if I keep these values? If not,
> I'll delete them.
>
No, it makes it less readable to continuously store and update these
values.

>

robertwb commented:
On Fri, Apr 13, 2018 at 2:23 PM Boyuan Zhang <notifications@github.com>
wrote:

> *@boyuanzz* commented on this pull request.
>
> *robertwb* wrote:
> OK, currently we have
>
> metricbase.Distribution
> metric.DelegatingDistribution
> metric.cells.DistributionCell
> metric.cells.DistributionResult
> metric.cells.DistributionData
> metric.cells.DistributionAggregator
> cy_combiners.DataflowDistributionCounter
> cy_combiners.DataflowDistributionCounterFn
> cy_dataflow_distribution_counter.DataflowDistributionCounter
>
> If you hadn't seen this code before (or came back in a while) how easy
> would it be to tell which one to use in which case. Which ones are fast?
> Where are the tests? How does this relate to the code in util/counters?
>
> I think this all needs to be consolidated and simplified (and that's
> probably beyond the scope of this PR, but this PR highlights the issue.
> Created https://issues.apache.org/jira/browse/BEAM-4074 which should
> probably be accompanied by a design doc).
>
> For now, let's stick with the convention cy_xxx/py_xxs or
> xxx_fast/xxx_slow for both the modules and the classes, and move everything
> out of cy_combiners. I'd also drop the "dataflow" modifier, but that's up
> to you.
>
> cc @pabloem <https://github.com/pabloem>
>
> I have two more questions here:
>
>    1. Is it okay to leave them still under transforms/ pkg at this point?
>
> Yes.

>
>    1. For current reporting routine design, I still need
>    DataflowDistributionCounterFn in cy_combiners, is it okay?
>
> Sure.

>
>
> For Dataflowmodifier, I would keep this at this point. Maybe it's better
> to drop when refactoring.
>
OK, if you can come up with a one sentence reason why these are Dataflow
Distribution Counters, not just Distribution Counters suitable for use
elsewhere. :)

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/5028#pullrequestreview-112163452>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAdqgQbPj0PVQ-_oMIRacX5PFBJFucjRks5toRc3gaJpZM4THsut>
> .
>

boyuanzz commented:
Latest commit https://github.com/apache/beam/pull/5028/commits/8111f084591f6e2bbb0063e7810334e4dd48661c addressed @robertwb 's comments.

boyuanzz commented:
Latest commit https://github.com/apache/beam/pull/5028/commits/1676f41a5a4a3aae55eea3cc0870c6c993121acc addressed @robertwb 's comments

boyuanzz commented:
Hey @robertwb @aaltay , do you think it's good time to squash all commits into one and waiting for merge?

robertwb commented:
Yes, please squash.

On Fri, Apr 13, 2018 at 4:32 PM Boyuan Zhang <notifications@github.com>
wrote:

> Hey @robertwb <https://github.com/robertwb> @aaltay
> <https://github.com/aaltay> , do you think it's good time to squash all
> commits into one and waiting for merge?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/5028#issuecomment-381284077>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAdqgXjVd-m6k6H1qkHrBOUXHSoGusUjks5toTWTgaJpZM4THsut>
> .
>

boyuanzz commented:
Squashed! Thanks so much for your patience. @robertwb
